### PR TITLE
pubsub: inject bundlers into NewTopic and NewSubscription

### LINKED ISSUE #665: pubsub: inject bundlers into NewTopic and NewSubscription
Currently these constructors take their own options structs and create bundlers behind the scenes, but they would be more flexible if they would accept bundlers as arguments instead. That would mean that the handlers these constructors currently hide as anonymous funcs would have to be made public funcs in the `pubsub` package, with comments saying that they are meant for driver use only.

### PR DESCRIPTION
**Pull request by ijt**
Fixes #665

### CODE REVIEW

File: internal/pubsub/driver/driver.go [hunk 1]
```diff
@@ -20,6 +20,12 @@ import (
 	"context"
 )
 
+// Batcher gathers items into batches to be sent to the pubsub service.
+type Batcher interface {
+	Add(ctx context.Context, item interface{}) error
+	Flush()
```
[the remark this answers is not in the document]
Comment (ijt) [reply internal/pubsub/driver/driver.go]: Done.

**Comment by vangent** [comment]
> we have no clear use case

The rationale for something like this is so that tests can use simpler and more predictable batchers. E.g., one test might want to use a batcher that it knows will batch up 3 messages and then send. Another test might want to use a no-op batcher that always immediately send everything. I predict it will be hard to write good tests with a dynamic batcher (and the tests will likely have to be rejiggered whenever the batcher is tuned).

That said, this does look more complicated than I thought it would be. I think it's the fact that the batcher needs a function to call when it wants to send a batch, and that function doesn't exist until you have the `Topic/Subscription`, making their definitions kind of circular.

Is it possible to make the batcher just given the `driver.SendBatch` function? I'm not sure that there needs to be any code in the concrete type about the batcher (other than calling its `Add` method).

**Comment by ijt** [comment]
@vangent, it seems straightforward to me. Can you be more specific about what's too complex?

I can remove the verbosity of the calls to `NewTopic` and `NewSubscription` by changing them back to their old signature and writing them in terms of new functions called `NewTopicWithBatcher` and `NewSubscriptionWithBatcher`.

I didn't understand your proposal. I'm up for an interactive chat if you have time.

**Comment by vangent** [comment]
It seems wrong to me that you're implementing the `Batcher` interface in the concrete type. The whole point of injecting it is that the batching algorithm is composable with the concrete type, and you should be able to use different `Batcher`s.

I'm thinking of something more like this:

```
// In the driver.
type Batcher interface {
  Add(ctx context.Context, item interface{}) error
  Shutdown()
}

type BatcherMaker func(itemExample interface{}, handler func(interface{}) error) Batcher

// In the concrete type.
func NewTopic(d driver.Topic, bm BatcherMaker) *Topic {
  handler := ... // existing code
  b := bm([]msgErrChan{},  handler)
  return &Topic{d, b}
}

// In the driver implementations.
func OpenTopic(...) (*pubsub.Topic, error) {
  // Public implementation uses the "real" batcher.
  bm := func(itemExample interface{}, handler func(interface{}) error) Batcher {
    return batcher.New(itemExample, 42, handler)
  }
  dt, err := openTopic(...) // returns driver.Topic
  // handle err
  return pubsub.NewTopic(dt, bm)
}

// In tests, use a different batcher.
  bm := func(_ interface{}, handler func(interface{}) error) Batcher {
    return MySimpleBatcher(handler) // e.g., just calls handler from Add
  }
  dt, err := openTopic(...) // returns driver.Topic
  t := pubsub.NewTopic(dt, bm)
}
```

Ping me on chat if you want to talk about it.

**Comment by ijt** [comment]
@vangent, probably most of the complexity here comes from the mismatch between the existing bundler and the `driver.Batcher` interface. @jba's new batcher implements the `driver.Batcher` interface with no changes, so once we start using that (see #689) the code here will be simpler.

**Comment by ijt** [comment]
@vangent, I think the change I made gives all the flexibility you're asking for, but it also includes some code to reduce call-site verbosity. The change here makes the batching algorithm composable with the concrete types of Topic and Subscription, or am I missing something?

**Comment by ijt** [comment]
From our chat, @vangent, I think I see what you're after now. We want to expose the handler funcs so tests can make simple batchers that use them.

I'm going to grab issue #689 now and come back to this change once that's done so this will be simpler.

**Comment by ijt** [comment]
The changes from #689 are integrated. PTAL

@vangent, I'd like to chat with you a bit more to better understand what you're hoping to see in this PR. The code you proposed has an interface return, something that's generally avoided: http://idiomaticgo.com/post/best-practice/accept-interfaces-return-structs/.

**Comment by vangent** [comment]
Per our discussion yesterday, let's leave this alone for now and see how far you get with the GCP record/replay tests.

The use case for this change is one where the record/replay tests are not repeatable due to race conditions. For example:

-- A test sends 5 messages concurrently.
-- The batcher will always send the 1st message in a batch of size 1, resulting in an RPC with a batch of size 1.
-- Depending on timing, the next batch could have 1, 2, 3, or 4 messages in it.
-- Thus, the test is flaky.

I'm not sure if we're going to need/want to write tests with this kind of concurrency, although it may be hard to avoid for `Acks` in particular.

We can discuss more in person tomorrow.

**Comment by ijt** [comment]
Okay, closing this out.

**Pull request closed without merging**